### PR TITLE
MDEV-30389: Update calls to `encryption_crypt` to provide correct destination length

### PR DESCRIPTION
**Pull request by tgross35**
- [x] *The Jira issue number for this PR is: [MDEV-30389](https://jira.mariadb.org/browse/MDEV-30389)*

## Description

As described in [MDEV-30389](https://jira.mariadb.org/browse/MDEV-30389), the value of `*dlen` passed to encryption plugins is uninitialized. It is an output variable so this is generally not a programmatic issue. However, there is currently no way to check the destination buffer length from the standpoint of an encryption plugin, so initializing `*dlen` to this value makes sense to allow for debug assertions and checked writes.

This PR seeks to update the passed value of `*dlen` to accurately reflect the available buffer size.

## How can this PR be tested?

This PR adds an `assert` statement to `include/mysql/service_encryption.h` that the destination buffer length is greater than or equal to the source buffer length. This will fail any calls in the existing test suite that do not pass a reasonable buffer length. (not 100% since the initial value is currently random and _may_ contain a value that passes, but so far it has proven "good enough" in practice)

## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

(could be either)

## Backward compatibility

`*dlen` is written by any encryption plugin, so it is exposed outside of MariaDB source. However, no plugin can be relying on the initial value of `*dlen` that this PR seeks to change, since it is currently uninitialized.

### CODE REVIEW

**Comment by tgross35** [comment]
`service_encryption.h` needs a `DBUG_ASSERT`, but I am not sure what to include to get the right one, since it is defined a couple places (and none seem to be in the include path for this header)

Needs update to `.pp` files, I'm also unsure how to do that

**Comment by grooverdan** [comment]
pp files are basicly a reference to ensure that we don't accidentality change the API.

**Comment by tgross35** [comment]
Thanks Dan, I might have messed something up but can't seem to get them to generate properly. I'll take a better look at it once I get tests passing for this patch.

**Comment by tgross35** [comment]
I think I have all calls updated to pass destination length equal to source length. Usually they are the same, but there are some cases where the destination buffer is longer. I'm working through updating these cases to be correct, but it's been a bit tricky to figure out.

But, I'll mark this as ready for review to get some general direction

File: storage/innobase/log/log0crypt.cc
Comment (dr-m) [inline storage/innobase/log/log0crypt.cc:226]: Please do not add C-style casts and do not replace TABs with spaces.

Comment (tgross35) [reply storage/innobase/log/log0crypt.cc:226]: Thanks, not sure why clang-tidy did that formatting

File: sql/encryption.cc
Comment (vuvova) [inline sql/encryption.cc:192]: I suppose it should be @invariant — everywhere else in the codebase doxygen commands start from `@`

File: include/mysql/service_encryption.h
Comment (vuvova) [inline include/mysql/service_encryption.h:121]: I'd also add something like
```
  assert(dst[*dlen-1]=1);
```
which _could_ crash and _will_ trigger asan/msan/valgrind if `*dlen` is too large

**Review by vuvova — CHANGES_REQUESTED** [review]
looks good. few comments, but nothing big.

File: include/mysql/service_encryption.h
Comment (tgross35) [reply include/mysql/service_encryption.h:44]: Thank you for the review, that explains the unexpected noise I was seeing when trying to update the .pp files.

I was actually looking to use `DBUG_ASSERT`, but if I recall correctly, I wasn't able to include `my_dbug.h` since it wasn't in the include path this file's build. If that sounds correct and there's a reason for it, do you know of a better solution? (It's been a while since I've looked at this, so I might just be thinking of the wrong thing)

Comment (vuvova) [reply include/mysql/service_encryption.h:44]: No, you remember correctly. `DBUG_ASSERT` is what the server normally uses for asserts. But this here is a header to be included by plugins, and I'd rather avoid getting `my_dbug.h` (and everything it needs in turn) into plugins. So to minimize plugin contamination by server headers, I would personally use `assert()` here.

Comment (tgross35) [reply include/mysql/service_encryption.h:44]: That makes sense, thank you for the explanation.

File: storage/innobase/log/log0crypt.cc
Comment (tgross35) [reply storage/innobase/log/log0crypt.cc]: I think this should be all set

**Comment by tgross35** [comment]
With just the length assert everything except autobake passes (f6f96b2df66662e431cee3b41de4132e1b4a0e7d), but adding the `assert((dst[*dlen - 1]= 1))` does make the binlog tests fail (ed8c9db2b2ee447808c96f7832908b027c37547c), so something needs some tweaking. Full failure list:

```
$ mysql-test/mtr --parallel=12 --mem --force --max-test-fail=400
...
Errors/warnings were found in logfiles during server shutdown after running the
following sequence(s) of tests:
   binlog_encryption.encrypted_slave
    binlog_encryption.encrypted_slave
    binlog_encryption.encrypted_slave
    binlog_encryption.multisource
    binlog_encryption.encryption_combo
    binlog_encryption.binlog_row_annotate
    binlog_encryption.binlog_row_annotate
    binlog_encryption.rpl_special_charset
    binlog_encryption.rpl_sporadic_master
    binlog_encryption.encrypted_slave
    binlog_encryption.rpl_sync
    binlog_encryption.encrypted_master
    binlog_encryption.encrypted_master
    binlog_encryption.rpl_parallel_analyze_table_hang
    binlog_encryption.rpl_init_slave_errors
    binlog_encryption.rpl_parallel_deadlock_corrupt_binlog
    binlog_encryption.rpl_loadfile
    binlog_encryption.rpl_parallel_domain
    binlog_encryption.rpl_mixed_binlog_max_cache_size
    binlog_encryption.rpl_parallel_domain_slave_single_grp
    binlog_encryption.rpl_switch_stm_row_mixed
    binlog_encryption.rpl_parallel_free_deferred_event
    binlog_encryption.binlog_incident
    binlog_encryption.rpl_parallel_incorrect_relay_pos
    binlog_encryption.rpl_loaddata_local
    binlog_encryption.rpl_parallel_innodb_lock_conflict
    binlog_encryption.rpl_ssl
    binlog_encryption.rpl_parallel_missed_error_handling
    binlog_encryption.rpl_special_charset
    binlog_encryption.rpl_parallel_mode
    binlog_encryption.rpl_sporadic_master
    binlog_encryption.rpl_parallel_record_gtid_wakeup
    binlog_encryption.rpl_sync
    binlog_encryption.rpl_parallel_retry_deadlock
    binlog_encryption.rpl_parallel_rollback_assert
    binlog_encryption.binlog_mdev_20574_old_binlog
    binlog_encryption.encrypted_slave
    binlog_encryption.encrypted_slave
    binlog_encryption.rpl_parallel_stop_on_con_kill
    binlog_encryption.binlog_xa_recover
    binlog_encryption.rpl_parallel_wrong_binlog_order
    binlog_encryption.encrypted_master
    binlog_encryption.encrypted_master
    binlog_encryption.rpl_relayrotate
    binlog_encryption.rpl_parallel_wrong_exec_master_pos
    binlog_encryption.rpl_init_slave_errors
    binlog_encryption.rpl_parallel_analyze_table_hang
    binlog_encryption.binlog_incident
    binlog_encryption.rpl_loadfile
    binlog_encryption.rpl_parallel_deadlock_corrupt_binlog
    binlog_encryption.rpl_loaddata_local
    binlog_encryption.rpl_switch_stm_row_mixed
    binlog_encryption.rpl_parallel_domain
    binlog_encryption.rpl_ssl
    binlog_encryption.rpl_parallel_domain_slave_single_grp
    binlog_encryption.rpl_semi_sync
    binlog_encryption.rpl_special_charset
    binlog_encryption.rpl_sporadic_master
    binlog_encryption.rpl_parallel_free_deferred_event
    binlog_encryption.rpl_sync
    binlog_encryption.rpl_stm_relay_ign_space
    binlog_encryption.rpl_parallel_incorrect_relay_pos
    binlog_encryption.rpl_parallel_gco_wait_kill
    binlog_encryption.rpl_parallel_innodb_lock_conflict
    binlog_encryption.rpl_parallel_gtid_slave_pos_update_fail
    binlog_encryption.encrypted_master
    binlog_encryption.rpl_parallel_missed_error_handling
    binlog_encryption.rpl_parallel_ignore_error_on_rotate
    binlog_encryption.rpl_parallel_mode
    binlog_encryption.rpl_parallel_ignored_errors
    binlog_encryption.encrypted_master
    binlog_encryption.rpl_parallel_record_gtid_wakeup
    binlog_encryption.rpl_parallel_partial_binlog_trans
    binlog_encryption.rpl_parallel_single_grpcmt
    binlog_encryption.rpl_parallel_retry_deadlock
    binlog_encryption.rpl_typeconv
    binlog_encryption.rpl_parallel_rollback_assert
    binlog_encryption.rpl_parallel_slave_bgc_kill
    binlog_encryption.rpl_parallel_analyze_table_hang
    binlog_encryption.rpl_parallel_stop_slave
    binlog_encryption.rpl_parallel_stop_on_con_kill
    binlog_encryption.rpl_init_slave_errors
    binlog_encryption.rpl_parallel_deadlock_corrupt_binlog
    binlog_encryption.rpl_parallel_wrong_binlog_order
    binlog_encryption.binlog_incident
    binlog_encryption.rpl_parallel_domain
    binlog_encryption.rpl_parallel_wrong_exec_master_pos
    binlog_encryption.rpl_loaddata_local
    binlog_encryption.rpl_parallel_domain_slave_single_grp
    binlog_encryption.rpl_ssl
    binlog_encryption.binlog_ioerr
    binlog_encryption.rpl_parallel_free_deferred_event
    binlog_encryption.mysqlbinlog
    binlog_encryption.rpl_parallel_incorrect_relay_pos
    binlog_encryption.mysqlbinlog
    binlog_encryption.rpl_parallel_innodb_lock_conflict
    binlog_encryption.rpl_semi_sync
    binlog_encryption.binlog_mysqlbinlog-cp932
    binlog_encryption.rpl_parallel_missed_error_handling
    binlog_encryption.binlog_index
    binlog_encryption.rpl_parallel_mode
    binlog_encryption.rpl_parallel_record_gtid_wakeup
    binlog_encryption.rpl_binlog_errors
    binlog_encryption.rpl_parallel_retry_deadlock
    binlog_encryption.rpl_binlog_errors
    binlog_encryption.rpl_parallel_rollback_assert
    binlog_encryption.rpl_parallel_stop_on_con_kill
    binlog_encryption.rpl_binlog_errors
    binlog_encryption.rpl_relayrotate
    binlog_encryption.rpl_parallel_wrong_binlog_order
    binlog_encryption.rpl_corruption
    binlog_encryption.rpl_corruption
    binlog_encryption.rpl_parallel_wrong_exec_master_pos
    binlog_encryption.rpl_corruption
    binlog_encryption.rpl_gtid_basic
    binlog_encryption.rpl_gtid_basic
    binlog_encryption.rpl_gtid_basic
    binlog_encryption.rpl_gtid_basic
    binlog_encryption.rpl_gtid_basic
    binlog_encryption.rpl_gtid_basic
    binlog_encryption.rpl_incident
    binlog_encryption.rpl_semi_sync
    binlog_encryption.rpl_incident
    binlog_encryption.rpl_incident
    binlog_encryption.rpl_packet
    binlog_encryption.rpl_parallel_show_binlog_events_purge_logs
    binlog_encryption.rpl_parallel_show_binlog_events_purge_logs
    binlog_encryption.rpl_parallel_show_binlog_events_purge_logs
    binlog_encryption.rpl_checksum
    binlog_encryption.rpl_temporal_format_default_to_default
    binlog_encryption.rpl_temporal_format_default_to_default
    binlog_encryption.rpl_relayrotate
    binlog_encryption.rpl_temporal_format_default_to_default
    binlog_encryption.rpl_temporal_format_mariadb53_to_mysql56
    binlog_encryption.rpl_temporal_format_mariadb53_to_mysql56
    binlog_encryption.rpl_skip_replication
    binlog_encryption.rpl_temporal_format_mysql56_to_mariadb53
    binlog_encryption.rpl_skip_replication
    binlog_encryption.rpl_skip_replication
    binlog_encryption.rpl_temporal_format_mysql56_to_mariadb53
    binlog_encryption.rpl_temporal_format_mysql56_to_mariadb53
    encryption.innodb_import
    encryption.innodb_import
198 tests were skipped, 90 by the test itself.
```

File: sql/log_event.cc
Comment (vuvova) [inline sql/log_event.cc:841]: this is why you got failures on `dst[*dstlen - 1]` assert.

here you should have `dstlen= sz - ev_offset - 4`

File: sql/log.cc
Comment (vuvova) [inline sql/log.cc:5665]: same here, must be `len - 4`

File: storage/innobase/fil/fil0crypt.cc
Comment (vuvova) [inline storage/innobase/fil/fil0crypt.cc:715]: this should be `uint` and after the `srclen` adjustment below

File: storage/innobase/fil/fil0crypt.cc
Comment (vuvova) [inline storage/innobase/fil/fil0crypt.cc:452]: this should be `uint`

File: storage/innobase/fil/fil0crypt.cc
Comment (vuvova) [inline storage/innobase/fil/fil0crypt.cc:713]: and here in the `encryption.innodb_import,encryption` test (when no page compression is used) encryption is called with `src == tmp` !
While we can adjust the assert to handle this, I think InnoDB shouldn't abuse the API like that.
@dr-m, what do you think?

Comment (tgross35) [reply storage/innobase/fil/fil0crypt.cc:713]: I found two calls that use the same source and destination buffer and changed them to allocate a temporary buffer instead https://github.com/MariaDB/server/pull/2438/commits/7237813317fe46256f0bcc2191a2e273acee6089. It doesn't look like the calls happen in any hot loops so the allocation overhead probably isn't too bad.

Also added an assertion to make sure they don't overlap and make the failure more explicit https://github.com/MariaDB/server/pull/2438/commits/7ea78661bc409556e986c912217805206f83950c. I'll adjust these changes as needed.

All tests pass on my local. It seems like there sometimes are a few spurious failures on CI, I'm not sure where they're coming from.

Comment (dr-m) [reply storage/innobase/fil/fil0crypt.cc:713]: @tgross35 Thank you. It does not look too bad, only affecting `ALTER TABLE…IMPORT TABLESPACE`. But could you please move the allocation of the temporary buffer up one level, to `fil_tablespace_iterate()`? That is where we allocate `iter.crypt_io_buffer` as well. Could we extend `fil_iterator_t` with a new field `crypt_tmp_buffer`?

**Review by vuvova — APPROVED** [review]
encryption API changes look good, thanks. As soon as Marko approves InnoDB new-buffer changes, we'll merge it.

File: storage/innobase/fil/fil0crypt.cc
Comment (tgross35) [reply storage/innobase/fil/fil0crypt.cc:713]: @dr-m I added a field to `fil_iterator_t` to help `FetchIndexRootPages::run` and updated the signature for `decrypt_decompress` to help `handle_instant_metadata`. I notice there don't seem to be any checks for alloc null/OOM - is this handled automatically somehow?

File: storage/innobase/log/log0crypt.cc
Comment (dr-m) [inline storage/innobase/log/log0crypt.cc:444]: Please do not use a C-style cast in C++ code. You have used `static_cast<uint>(size)` elsewhere.

File: storage/innobase/fil/fil0crypt.cc
Comment (dr-m) [reply storage/innobase/fil/fil0crypt.cc:713]: For `aligned_malloc()` some explicit OOM handling could indeed be added fairly easily. InnoDB used to be very unforgiving with regard to memory allocations, killing the server whenever running out of memory.

**Review by dr-m — CHANGES_REQUESTED** [review]

File: storage/innobase/row/row0import.cc
Comment (tgross35) [reply storage/innobase/row/row0import.cc:3129]: Thanks for the review.

I didn't think this needs to be aligned at all so I just meant to be consistent with the other allocs in that section. I figured that maybe anything largish just gets page alignment to prevent some fragmentation via allocation offsets.

I'll update this.

Comment (tgross35) [reply storage/innobase/row/row0import.cc:3129]: Is there an advantage of aligned_malloc to `CPU_LEVEL1_DCACHE_LINESIZE` vs. just `malloc` in this case?

**Comment by tgross35** [comment]
Thanks for the review @dr-m. Regarding formatting - can clang-format always be relied on? It seemed to make some unneeded changes (like here https://github.com/MariaDB/server/pull/2438#discussion_r1073212389) and I was surprised it wasn't run as part of CI.

**Comment by tgross35** [comment]
I believe all the concerns have been addressed, so I marked this as ready for review.

**Comment by tgross35** [comment]
The windows build failure looks unrelated, if anyone has the ability to rebuild.

**Comment by tgross35** [comment]
@dr-m gentle poke, would you be able to re-review this?

I will need to rebase before it can be merged, but I don't see any conflicts.

File: storage/innobase/include/log0crypt.h
Comment (dr-m) [inline storage/innobase/include/log0crypt.h:88]: We do not use backticks when referring to variables or parameters in comments. "Length" is misspelled, and the indentation was broken here.

File: storage/innobase/log/log0crypt.cc
Comment (dr-m) [inline storage/innobase/log/log0crypt.cc:432]: Do not use ' or ` in comments that refer to parameters, and do not break the indentation.

File: storage/innobase/row/row0import.cc
Comment (dr-m) [inline storage/innobase/row/row0import.cc:4180]: This would fit in a single line if you wrote `* srv_page_size`. The rest of the code is already referring to `srv_page_size`, so it could save a load of another memory location. Most CPUs have instructions for fast integer multiplication.

File: storage/innobase/row/row0import.cc
Comment (dr-m) [inline storage/innobase/row/row0import.cc:4209]: It is awkward, but there are two formatting styles in InnoDB code base. The old style (as in this function) uses TAB for indentation, K&R style placement of opening braces, and a space on both sides of assignment operators. The two formatting styles should not be mixed, and unnecessary changes to formatting should be avoided.

Unfortunately, the code base is not compatible with tools like `clang-format`.

**Review by dr-m — CHANGES_REQUESTED** [review]
I am sorry for missing the notification.

File: storage/innobase/include/log0crypt.h
Comment (tgross35) [reply storage/innobase/include/log0crypt.h]: This should be resolved

File: storage/innobase/log/log0crypt.cc
Comment (tgross35) [reply storage/innobase/log/log0crypt.cc]: This should be resolved

File: storage/innobase/row/row0import.cc
Comment (tgross35) [reply storage/innobase/row/row0import.cc]: This was just copied from what was there before, but I updated it.

File: storage/innobase/include/log0crypt.h
Comment (tgross35) [reply storage/innobase/include/log0crypt.h]: This should be resolved

File: storage/innobase/row/row0import.cc
Comment (tgross35) [reply storage/innobase/row/row0import.cc]: I believe I have corrected this now, but could use another look.

Perhaps you are already aware but if not - there's a recent feature in github where you can add a `.git-blame-ignore-revs` file and it will [hide those commits from the blame](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view). Useful for doing a one time whole-codebase formatting commit so you can enforce style in CI. Not sure how other tooling might interact, but it could be a path to reduce reviewer & editor burden of mixed styles in the future.

Anyway, I think I covered the necessary changes and have requested a re-review.

**Comment by tgross35** [comment]
This _should_ all be good but it looks like one random test failed - would anyone be able to retry that step?

**Comment by vuvova** [comment]
That test failure clearly has nothing to do with encryption. I'll wait for @dr-m to approve and then we'll merge it. Thanks!

**Comment by dr-m** [comment]
> That test failure clearly has nothing to do with encryption. I'll wait for @dr-m to approve and then we'll merge it. Thanks!

The failure due to result difference for the test `innodb.innodb-index-online` looks timing-related. MemorySanitizer can introduce 250% of run-time overhead. The test does make use of encryption, and the failed test combination does involve encryption, so it could theoretically be related to this change. However, similar failures have occurred in the past.

File: storage/innobase/row/row0import.cc
Comment (dr-m) [inline storage/innobase/row/row0import.cc:242]: It’d be better not to reformat the pre-existing code lines.

File: storage/innobase/row/row0import.cc
Comment (dr-m) [inline storage/innobase/row/row0import.cc:3136]: I think that the existing formatting was according to the default settings of GNU Emacs’ `c++-mode`.

File: storage/innobase/row/row0import.cc
Comment (dr-m) [inline storage/innobase/row/row0import.cc:3189]: Same here: the contents of the `{}` block should not be deindented.

File: storage/innobase/row/row0import.cc
Comment (dr-m) [inline storage/innobase/row/row0import.cc:3272]: Ditto

**Review by dr-m — APPROVED** [review]
Thank you, this looks better.

  I think that we should re-evaluate the prospect of configuring and applying `clang-format` to at least some of the code base. I was not aware of the existence of the `git blame --ignore-revs-file` or `blame.ignoreRevsFile` options.

**Comment by tgross35** [comment]
I fixed the formatting issues, but the msan test is still failing - I'm not very familiar with these tests, is there a way that timeout or output should be adjusted here?

```
innodb.innodb-index-online '4k,clear,innodb' w28 [ fail ]
        Test ended at 2023-06-20 05:07:44

CURRENT_TEST: innodb.innodb-index-online
--- /home/buildbot/amd64-debian-11-msan/build/mysql-test/suite/innodb/r/innodb-index-online.result	2023-06-20 04:56:32.000000000 +0000
+++ /home/buildbot/amd64-debian-11-msan/build/mysql-test/suite/innodb/r/innodb-index-online.reject	2023-06-20 05:07:43.742274058 +0000
@@ -170,8 +170,8 @@
 ERROR 70100: Query execution was interrupted
 SELECT name, count FROM INFORMATION_SCHEMA.INNODB_METRICS WHERE subsystem = 'ddl';
 name	count
-ddl_background_drop_indexes	0
-ddl_online_create_index	0
+ddl_background_drop_indexes	1
+ddl_online_create_index	1
 ddl_pending_alter_table	0
 ddl_sort_file_alter_table	0
 ddl_log_file_alter_table	0

mysqltest: Result content mismatch
```

`main.group_min_max` is also failing after rebase, but this also fails on the 11.0 branch.

**Comment by dr-m** [comment]
Thank you, @tgross35. The MSAN failure was for an `innodb_page_size=4k` variant of `innodb.innodb-index-online,clear` where no encryption is used, so it should be unrelated to this.

For some reason, I can’t update this to the latest base branch for final merging.

**Comment by vuvova** [comment]
I'll merge it

**Comment by tgross35** [comment]
Thank you!

**Comment by dr-m** [comment]
Thank you for your patience.

**Pull request closed without merging**
